File: csharp/patterns/src/PrototypePatterns/Behavioral/State/StatePattern.cs
namespace PrototypePatterns.Behavioral.State;

public interface IOrderState
{
    string Advance(OrderContext context);
}

public sealed class PendingState : IOrderState
{
    public string Advance(OrderContext context)
    {
        context.SetState(new PaidState());
        return "pending->paid";
    }
}

public sealed class PaidState : IOrderState
{
    public string Advance(OrderContext context)
    {
        context.SetState(new ShippedState());
        return "paid->shipped";
    }
}

public sealed class ShippedState : IOrderState
{
    public string Advance(OrderContext context) => "shipped->shipped";
}

public sealed class OrderContext
{
    private IOrderState _state = new PendingState();

    public void SetState(IOrderState state) => _state = state;

    public string Advance() => _state.Advance(this);
}

public static class StatePattern
{
    public static IReadOnlyList<string> Demo()
    {
        var order = new OrderContext();
        return [order.Advance(), order.Advance(), order.Advance()];
    }
}

File: java/patterns/src/main/java/org/prototypepatterns/behavioral/state/StatePattern.java
package org.prototypepatterns.behavioral.state;

import java.util.List;

public final class StatePattern {
    private StatePattern() {
    }

    interface OrderState {
        String advance(OrderContext context);
    }

    static final class PendingState implements OrderState {
        @Override
        public String advance(OrderContext context) {
            context.setState(new PaidState());
            return "pending->paid";
        }
    }

    static final class PaidState implements OrderState {
        @Override
        public String advance(OrderContext context) {
            context.setState(new ShippedState());
            return "paid->shipped";
        }
    }

    static final class ShippedState implements OrderState {
        @Override
        public String advance(OrderContext context) {
            return "shipped->shipped";
        }
    }

    static final class OrderContext {
        private OrderState state = new PendingState();

        void setState(OrderState state) {
            this.state = state;
        }

        String advance() {
            return state.advance(this);
        }
    }

    public static List<String> demo() {
        OrderContext context = new OrderContext();
        return List.of(context.advance(), context.advance(), context.advance());
    }
}
